crypto/rand/md_rand.c

```diff
@@ -194,16 +191,41 @@ RAND_METHOD *RAND_SSLeay(void)
     return (&rand_ssleay_meth);
 }
 
+static void ssleay_rand_free(MD_RAND *r)
+{
+    OPENSSL_free(r);
+}
+
+static MD_RAND *ssleay_rand_get(void)
+{
+    MD_RAND *state = CRYPTO_get_thread_local(CRYPTO_THREAD_LOCAL_RAND);
+    if (state == NULL) {
+        state = OPENSSL_zalloc(sizeof(*state));
+        if (state == NULL)
+            return NULL;
+
+        if (!CRYPTO_set_thread_local(CRYPTO_THREAD_LOCAL_RAND, state,
+                                     (void (*)(void *))ssleay_rand_free)) {
+            OPENSSL_free(state);
+            return NULL;
+        }
+    }
+
+    return state;
+}
+
 static void ssleay_rand_cleanup(void)
 {
-    OPENSSL_cleanse(state, sizeof(state));
-    state_num = 0;
-    state_index = 0;
-    OPENSSL_cleanse(md, MD_DIGEST_LENGTH);
-    md_count[0] = 0;
-    md_count[1] = 0;
-    entropy = 0;
-    initialized = 0;
+    MD_RAND *r = ssleay_rand_get();
+
+    OPENSSL_cleanse(r->state, sizeof(r->state));
```

Comment:
So now RAND_cleanup only affects the current thread?  That doesn't seem like the intended behavior.